pkg/datastore/postgres/repositories/checked_headers_repository.go

```diff
@@ -17,62 +17,77 @@
 )
 
 const (
-	insertCheckedHeaderQuery = `UPDATE public.headers SET check_count = (SELECT check_count WHERE id = $1) + 1 WHERE id = $1`
+	insertCheckedHeaderQuery = `
+INSERT INTO %s.checked_headers (check_count, header_id)
+VALUES (1, $1)
+ON CONFLICT (header_id) DO
+	UPDATE SET check_count =
+		(SELECT checked_headers.check_count WHERE checked_headers.header_id = $1) + 1
+	WHERE checked_headers.header_id = $1`
 )
 
 type CheckedHeadersRepository struct {
-	db *postgres.DB
+	db         *postgres.DB
+	schemaName string
 }
 
-func NewCheckedHeadersRepository(db *postgres.DB) CheckedHeadersRepository {
-	return CheckedHeadersRepository{db: db}
+func NewCheckedHeadersRepository(db *postgres.DB, schemaName string) CheckedHeadersRepository {
+	return CheckedHeadersRepository{db: db, schemaName: schemaName}
 }
 
 // Increment check_count for header
 func (repo CheckedHeadersRepository) MarkHeaderChecked(headerID int64) error {
-	_, err := repo.db.Exec(insertCheckedHeaderQuery, headerID)
+	queryString := fmt.Sprintf(insertCheckedHeaderQuery, repo.schemaName)
+	_, err := repo.db.Exec(queryString, headerID)
 	return err
 }
 
 // Zero out check count for header with the given block number
 func (repo CheckedHeadersRepository) MarkSingleHeaderUnchecked(blockNumber int64) error {
-	_, err := repo.db.Exec(`UPDATE public.headers SET check_count = 0 WHERE block_number = $1`, blockNumber)
+	queryString := fmt.Sprintf(`UPDATE %s.checked_headers ch
+								SET check_count = 0
+								FROM public.headers h
+								WHERE ch.header_id = h.id
+								AND h.block_number = $1`, repo.schemaName)
+
+	_, err := repo.db.Exec(queryString, blockNumber)
 	return err
 }
 
 // Return header if check_count  < passed checkCount
 func (repo CheckedHeadersRepository) UncheckedHeaders(startingBlockNumber, endingBlockNumber, checkCount int64) ([]core.Header, error) {
 	var (
 		result                  []core.Header
-		query                   string
 		err                     error
 		recheckOffsetMultiplier = 15
 	)
 
+	joinQuery := fmt.Sprintf(`SELECT h.id, h.block_number, h.hash
+						FROM public.headers h
+						LEFT JOIN %s.checked_headers ch
+						ON ch.header_id = h.id`, repo.schemaName)
+
 	if endingBlockNumber == -1 {
-		query = `SELECT id, block_number, hash
-			FROM public.headers
-			WHERE (check_count < 1
-			           AND block_number >= $1)
-			   OR (check_count < $2
-			           AND block_number <= ((SELECT MAX(block_number) FROM public.headers) - ($3 * check_count * (check_count + 1) / 2)))`
-		err = repo.db.Select(&result, query, startingBlockNumber, checkCount, recheckOffsetMultiplier)
+		noEndingBlockQuery := fmt.Sprintf(`%s
+				 WHERE ((ch.check_count IS NULL OR ch.check_count < 1) AND h.block_number >= $1)
+				 OR ((ch.check_count IS NULL OR ch.check_count < $2)
+			           AND h.block_number <= ((SELECT MAX(block_number) FROM public.headers) - ($3 * ch.check_count * (ch.check_count + 1) / 2)))`, joinQuery)
```

Comment:
So seems like this may have been the case in the earlier implementation too, but does the second branch of this where clause disregard the starting block number?

Reply:
Looks like it does to me too. I hadn't noticed by I believe that's an error (albeit not a new one). Can try it out with a test.

Reply:
I cleaned up this query a bit while fixing this bug, but I fixed it and added some tests.